sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -194,46 +167,36 @@ class CacheManager extends Logging {
   private def recacheByCondition(
       spark: SparkSession,
       condition: CachedData => Boolean): Unit = {
-    val needToRecache = scala.collection.mutable.ArrayBuffer.empty[CachedData]
-    readLock {
-      val it = cachedData.iterator()
-      while (it.hasNext) {
-        val cd = it.next()
-        if (condition(cd)) {
-          needToRecache += cd
-        }
-      }
+    val needToRecache = cachedData.filter(condition)
+    this.synchronized {
+      // Remove the cache entry before creating a new ones.
+      cachedData = cachedData.filterNot(cd => needToRecache.exists(_ eq cd))
```

Comment:
I talked with @maropu. Since write to a `volatile` reference is [atomic](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.7), I think that we could remove `this.sychronization` if we would allow this operation `cachedData.filterNot(cd => needToRecache.exists(_ eq cd))` to run on multiple threads.

Since the performance bottleneck was in read lock, I do not recommend to apply this change very soon.

Reply:
The write itself is atomic, but you need the this.synchronized to ensure that the value you are writing is using the most up to date value of cachedData.  The value is read as part of cachedData.filterNot, and you need to make sure cachedData doesn't change between the time when it is read and when it is written.